File: y-comb/yComb.ts
type Rec<A> = (rec: Rec<A>) => A;

function selfApp<A>(f: (a: () => A) => A): A {
  const selfAppInternal: Rec<A> = (rec: Rec<A>) => f(()=> rec(rec))
  return selfAppInternal(selfAppInternal);
}

let maxNat: (m: number, n: number) => number
  = selfApp(rec => (m, n) => m === 0 ? n :
                             n === 0 ? m : 1 + rec()(m-1,n-1));

let factorial: (n: number) => number
  = selfApp(rec => n => n === 0 ? 1 : n * rec()(n-1));

console.log(maxNat(10,6)); // 10
console.log(factorial(5)); // 120

File: y-comb/yComb.hs
newtype Rec a = Rec {unwrap :: (Rec a) -> a}

selfApp :: (a -> a) -> a
selfApp f =
  let selfAppInternal = \rec -> f ((unwrap rec) rec)
  in selfAppInternal (Rec selfAppInternal)

maxNat :: Integer -> Integer -> Integer
maxNat = selfApp(\rec -> \m n -> if m == 0 then n else
                                 if n == 0 then m else 1 + (rec (m - 1) (n - 1)))

factorial :: Integer -> Integer
factorial = selfApp(\rec -> \n -> if n == 0 then 1 else n * (rec (n - 1)))
